src/modules/googlesheets.ts

```diff
@@ -0,0 +1,149 @@
+import axios from 'axios';
+import _ from 'lodash';
+import { AccessInfoBigQueryModel } from '../generated/tdr';
+import { showNotification } from './notifications';
+
+export type SpreadsheetInfo = {
+  spreadsheetId: string;
+  spreadsheetUrl: string;
+};
+
+export type SheetInfo = {
+  sheetId: string;
+  title: string;
+};
+
+export const createSheet: any = async (sheetName: string, token: string) => {
+  // Using google drive api endpoint rather than google sheets api
+  // in order to facilitate file delete on error
+  const url = '/drive/v3/files';
+  return axios
+    .post(
+      url,
+      { mimeType: 'application/vnd.google-apps.spreadsheet', name: sheetName },
+      {
+        headers: {
+          'Content-Type': 'application/json',
+          Authorization: `Bearer ${token}`,
+        },
+      },
+    )
+    .then((response) => ({
+      spreadsheetId: response.data.id,
+      spreadsheetUrl: `https://docs.google.com/spreadsheets/d/${response.data.id}`,
+    }))
+    .catch((err) => {
+      showNotification(err);
+    });
+};
+
+// Adding data sources for each table one by one
+// In theory, could do a batch request for all data sources
+// But ran into errors around concurrent requests from the sheets api
+export const addBQSources: any = async (
+  spreadsheetId: string,
+  bigQueryAccessInfo: AccessInfoBigQueryModel,
+  token: string,
+) => {
+  const url = `/googlesheets/v4/spreadsheets/${spreadsheetId}:batchUpdate`;
+  const sheetInfo: Array<SheetInfo> = [];
+  let index = 0;
+  for (const table of bigQueryAccessInfo.tables) {
+    const requests: object[] = [
+      {
+        addDataSource: {
+          dataSource: {
+            spec: {
+              bigQuery: {
+                projectId: bigQueryAccessInfo.projectId,
+                querySpec: {
+                  rawQuery: `select * from \`${bigQueryAccessInfo.projectId}.${bigQueryAccessInfo.datasetName}.${table.name}\``,
```

Comment:
Putting my HackEDU hat on, I'm concerned by the raw SQL query here. Even though it's not based on user input, someone could theoretically intercept and modify the API response right?

Could we add some sanitization here? We could also add ANTLR so we’re doing the same query parsing as our API, but that's probably worth splitting out into a separate ticket.

Reply:
Thanks for putting on your HackEDU Hat 😄 
The main reason why we don't have to be too concerned here is that the scope defined here is only for read-only access to big query. 

If still want to avoid building a query in the UI (which is totally fair), then we could use "sampleQuery" provided by the getDataset/snapshot endpoints. I just have to remove the "LIMIT 1000" clause so the sheet has access to the whole dataset. I just put up a draft PR and would appreciate any thoughts on that approach! https://github.com/DataBiosphere/jade-data-repo/pull/1286

Reply:
That's reassuring! If it's not a big of a lift I think using the sampleQuery would be great